Draft/add library

pengale opened the pull request.

Commit:
Initial commit of a the rollingops framework.

This is an example charm showcasing an implementation of the Rolling
Operations library.

The library is included with the example charm.

This is a first take on an asynchronous lock built on top of the
primitives in the Operator Framework. It passes unit tests, but has
not been extensively tested in a live charm.

rwcarlsen commented on src/charm.py:
```diff
@@ -9,28 +9,32 @@
+
+        self.restart = RollingOpsManager(self, 'restart', self._restart, self.on.restart_action)
+
+        # Sentinal for testing (omit from production charms)
```
"Sentinel" :-)

pengale replied:
Whoops! Thx!

rwcarlsen commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+            return
+
+        if Lock(self.name, self.charm).is_held():
+            self.charm.on.run_with_lock.emit(name=self.name)
```
What if this hangs or has problems?  Possibly want to think about exception/error handling and what that does to the rolling process.

pengale replied:
@rwcarlsen if it hangs, the whole process is blocked, and a human operator will need to intervene and figure out what went wrong with the unit that hung. That workflow is intentional, though I managed to leave out the part where the executing charm sets a "executing {}".format(self.name)" status so that this is a bit more discoverable :-)

(Also, I should add some comments to that effect.)

We do need a way to release the lock for a unit that goes away, possibly automagically. I'll work on that. (We could also add a timestamp to the locks, which would give us a way of discarding old lock requests.)

rwcarlsen commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+            # This is not relevant to us.
+            return
+
+        if Lock(self.name, self.charm).is_held():
+            self.charm.on.run_with_lock.emit(name=self.name)
+
+        if self.framework.model.unit.is_leader():
+            self.charm.on.process_locks.emit(name=self.name)
```
I'm still not convinced that these custom events are helpful. Do we anticipate a charm code outside of the rolling class responding-to/listening-for these events? If not, then I think the code would be clearer if RollingOpsManager just called its run_with_lock, etc. functions directly.

pengale replied:
@rwcarlsen I went back and forth on this. `self.charm.on.process_locks.emit` is definitely wordier than "self._process_locks", and they both are equally easy to override if extending this library.

But the event contains the following logic for free:

```
try:
    self._on_process_locks()
except Exception:
   self._store.run_with_lock = "already_executed"
   event.defer()
   return
else:
   self._store.run_with_lock = ""  # Clear for next time.
```

I like that a lot. Driving things as events provides a nice separation of concerns, and some clarity in the interface.

It would be a *lot* cleaner if the custom events contained their own handlers, which got automagically attached to the charm, but I'm not sure that this is the right place to establish that pattern.

pengale replied:
Following up on this: TIL that you can run custom events from the juju cli!

That has the potential to make all of this a lot cleaner. You can drop the rigamarole with defining and action, and just expose an event that operators (both human and software) can call on when they want to. I like this a whole lot :-)

rwcarlsen commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+            lock.grant()
+            if lock.unit == self.charm.model.unit:
+                # It's time for the leader to run with lock.
+                self.charm.on.run_with_lock.emit(name=self.name)
```
Is a relation_changed event not fired for a unit when it modifies its own relation data?

pengale replied:
In the test harness, it does not.

In production, it might -- that's one of the things that I need to test before this becomes production worthy.

rwcarlsen commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+        self.framework.observe(charm.on.run_with_lock, self._on_run_with_lock)
+        self.framework.observe(charm.on.process_locks, self._on_process_locks)
+
+    def _func_run(self: CharmBase, event: EventBase) -> None:
+        """Placeholder for the function that actually runs our event.
+
+        Usually overridden in the init.
+        """
+        raise NotImplementedError
```
Not sure what this is doing here.  Is it a convention to make placeholder methods like this that get overridden with func values/lambdas?

pengale replied:
It's a convention which we've used elsewhere at Canonical (though I may have picked it up earlier than that). 

It helps the class as written better reflect the class at runtime, it drops something into any autogenerated documentation, and it gives folks another chance to wrap their head around the idea that this won't be functional unless they pass an executable in.

marcoppenheimer replied:
I'm a fan of it!

rwcarlsen commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+
+        # If we reach this point, and we have pending units, we want to grant a lock to
+        # one of them.
+        if pending:
```
Would it be useful having some kind of "rolling process finished" func that the charm also passes in that we call after all the locks have been processed (e.g. called in an else condition for this if)?

pengale replied:
As written, we don't have a way of distinguishing old and new locks. A lock starts in "idle" and ends in "idle". The nice thing about that is that we don't need any logic to handle the case where a human operator gets impatient and sends multiple requests to kick off the restart.

The not nice thing is that it's harder to send an "all done!" message, and also harder to trigger the start of a rolling restart without running an action.

I think that we might be heading toward timestamps. I just wanted to start simple, just in case we could get away with simple :-)

jnsgruk replied:
Heh. I just suggested the same thing above -- I think the juju-y way to handle this would be to emit an event signalling that all units have held the lock and run the closure -- then authors could decide (or not) to subscribe to it and respond appropriately in their charms.

rwcarlsen commented on src/charm.py:
```diff
@@ -9,28 +9,32 @@
 class CharmRollingOpsCharm(CharmBase):
     """Charm the service."""
 
+    on = RollingEvents()
```
Is this really required here?  It seems like maybe not since the charm itself never binds to any of the rolling events - those are all only used internally by the RollingOpsManager.

pengale replied:
We do bind stuff to the charm (e.g. charm.on.process_locks). And I don't think that the Framework will pay attention to events that haven't been added to the Charm's offical list 'o events.

I can dig into the internals more, though. I would love to omit this line, especially if I could do so in a way that namespaced all these things.

rwcarlsen replied:
In my mind - it seems like there should be a way for the ops manager class to observe necessary events (including custom) without having to modify the on member of the charm class.  If there isn't a way, we should make a way.

jnsgruk replied:
Agree with Robert - there is a way I noted in a comment above :)

pengale replied:
One thing that I want to do, though I still have to do some more double checking and testing, is namespace all of this. 

So you bind to on['namespace'].a_library_event

That should get us to a place where we aren't clobbering things.

Event with namespacing, there is a lot of benefit to having to explicitly modify "on" in your own code. When you do so, you are explicitly verifying that you trust the library to add events. That's nice because it gives charm authors a moment to pause and decide whether or not they trust the library. And it also makes it clear that this library is changing some of the fundamental behavior of the charm ... meaning that you can take advantage of this fact, and bind your own handlers to the library events!

rwcarlsen reviewed (commented):
Overall it looks pretty good.  I really like the minimal API for the charm - basically one line creating the rolling ops manager.

Commit:
Fixed action params.

I had the wrong idea about how the params actually get attached to an
ActionEvent.

Revised Mock and code reflect actual production behavior.

marcoppenheimer commented:
I'm really impressed with it, especially the abstraction from only restarts to generic `func_run`.

One thing that wasn't entirely clear reading through was the `event_acquire` arg to RollingOpsManager.

In the example you gave, `on.restart_action` is passed, which is used as `self.framework.observe(event_acquire, self._on_acquire_lock)` and is never seen again. More for my understanding, could you please walk through what it's doing?

marcoppenheimer commented on metadata.yaml:
```diff
@@ -15,3 +15,7 @@ bases:
       - amd64
+peers:
+    restart:
+        interface: rolling_op
```
This is missing an `s`?

rwcarlsen commented:
I imagined this sequence that could break the current implementation:

* A rolling restart action is half done
* unit 7 just finishes its restart and releases the lock (modifies relation data)
* simultaneously: juju sends out relation changed events, juju changes leader from unit 42 to unit 999
* unit 999 ignores relation-changed because it's not leader (received relation changed event first)
* unit 42 ignores relation-changed because it's not the leader (received leader change state/info first)

The rolling restart process is now stalled.  Everybody ignored the relation changed events that were suppose to trigger the next ops manager lock wrangling.  I'm not completely sure this could happen, but it seems plausible.

pengale commented:
> I'm really impressed with it, especially the abstraction from only restarts to generic `func_run`.
> 
> One thing that wasn't entirely clear reading through was the `event_acquire` arg to RollingOpsManager.
> 
> In the example you gave, `on.restart_action` is passed, which is used as `self.framework.observe(event_acquire, self._on_acquire_lock)` and is never seen again. More for my understanding, could you please walk through what it's doing?

Thank you!

And I will add some docs around the "restart action". The summary goes something like this:

When we define the action in actions.yaml, we make our charm capable of receiving an event called "action_<action_name>". We don't need to define that anywhere in code. It's just "if there exists an action called foo in this charm, the charm can list for an event called action_foo, which fires when an operator runs the action against this unit.

We won't always want to trigger a restart based on an action. But we can, by telling the library to react based on charm.on.action_<name>. (I decided to pass it in to give a charm author some flexibility.)

Commit:
Added basic integration tests.

And fixed some issues that I discovered!

The tests use unittest.IsolatedAsyncioTestCase.

We now create a model, so we can do a fresh deploy internal to the test.

Also .charm file to .gitignore.

pengale commented:
> I imagined this sequence that could break the current implementation:
> 
>     * A rolling restart action is half done
> 
>     * unit 7 just finishes its restart and releases the lock (modifies relation data)
> 
>     * simultaneously: juju sends out relation changed events, juju changes leader from unit 42 to unit 999
> 
>     * unit 999 ignores relation-changed because it's not leader (received relation changed event first)
> 
>     * unit 42 ignores relation-changed because it's not the leader (received leader change state/info first)
> 
> 
> The rolling restart process is now stalled. Everybody ignored the relation changed events that were suppose to trigger the next ops manager lock wrangling. I'm not completely sure this could happen, but it seems plausible.

This sounds plausible.

I think that we can fix it by emitting a "process_locks" event on leader-elected. And also possibly tie it to update_status, though I'd like to avoid the latter if we can.

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+
+class SomeCharm(...):
+    # ...
+    on = RollingEvents
```
I feel like we can take care of this for the developer in the library init? We're passing `self` into the library, so the library could do a `self.charm.on = RollingEvents()`. My only concern is if that might override any *other* custom event sources, so perhaps `self.charm.on['rolling_ops'] = RollingEvents()` to keep them 'namespaced'?

Also I think it should be `RollingEvents()` not `RollingEvents`, right?

pengale replied:
The risk of silently clobbering is why I left this here (as per our docs), rather than trying to fold it into the library.

And it is RollingEvents, with no call. The class is basically just a list of allowed events. Initializing it doesn't do anything.

As we use more libraries, we need a better way of doing this, but I'm not ready to tackle that quite yet. For now, this at least makes it clear that you will be creating problems if you try to attach the events from two different libraries to on.

pengale replied:
Edit: nevermind. The call to init the class is necessary. Per my own code, written below :-)

But it is Events, plural, not Event, singular.

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+
+
+class LockErrorNoRelation(Exception):
+    pass
```
This could do with a docstring explaining when/why this exception would be raised :)

pengale replied:
It gets raised when there is no relation. But yes, I can add a docstring :-)

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+    This class tracks state across the cloud by implementing a peer relation
+    interface. There are two parts to the interface:
+
+    1) The data on a unit's peer relation (defined in metadata.yaml.) Each unit can update
+       this data. The only meaningful values are "acquire", and "release", which represent
+       a request to acquire the lock, and a request to release the lock, respectively.
+
+    2) The application data in the relation. This tracks whether the lock has been
+       "granted", Or has been released (and reverted to idle). There are two valid states:
+       "granted" or None.  If a lock is in the "granted" state, a unit should emit a
+       RunWithLocks event and then release the lock.
+
+       If a lock is in "None", this means that a unit has not yet requested the lock, or
+       that the request has been completed.
+
+    In more detail, here is the relation structure:
+
+    relation.data:
+        <unit n>:
+            status: 'acquire|release'
+        <application>:
+           <unit n>: 'granted|None'
+
+    Note that this class makes no attempts to timestamp the locks and thus handle multiple
+    requests in a row. If a unit re-requests a lock before being granted the lock, the
+    lock will simply stay in the "acquire" state. If a unit wishes to clear its lock, it
+    simply needs to call lock.release().
```
This is nicely articulated, thanks :)

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+
+        if app_status == self.GRANTED and unit_status == self.RELEASE:
+            # Active release request.
+            return self.RELEASE
```
I wonder if these consts could in fact be a pair of Enums, such that you could decorate this method with a more concrete return type?

pengale replied:
Yep. We can do that.

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+        self.app = charm.framework.model.app
+
+    @property
+    def status(self):
```
Would be nice to see a type annotation here (and elsewhere in the public API at least), or some more description of the return types/values in the docstring

pengale replied:
Makes sense. I will make the change.

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+
+        Since we update the relation data, this may fire off a RelationChanged event.
+        """
+        if status == self.ACQUIRE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.RELEASE:
+            self.relation.data[self.unit].update({"status": status})
```
```suggestion
        if status == self.ACQUIRE or status == self.RELEASE:
            self.relation.data[self.unit].update({"status": status})
```

pengale replied:
+1 (I've got local changes, so I'm not going to merge just yet.)

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+        if status == self.RELEASE:
+            self.relation.data[self.unit].update({"status": status})
+
+        if status == self.GRANTED:
+            self.relation.data[self.app].update({self.unit: status})
+
+        if status is self.IDLE:
+            self.relation.data[self.app].update({self.unit: status})
```
```suggestion
        if status == self.GRANTED or status == self.IDLE:
            self.relation.data[self.app].update({self.unit: status})
```

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+
+    def __init__(...)
+        # ...
+        self.restart = RollingOpsManager(self, 'restart', self._restart, self.on.restart_action)
```
I wonder if we should make the args kwarg only here? I feel like it'll improve the readability of charms quite a bit?

pengale replied:
We could. I initially had coded this as keyworded args, and got rid of them for brevity's sake.

I really would like to chop out at least two more of the args, but I'm not 100% certain how to do so.

(It's possible that the way we resolved multiple sets of events from multiple libraries will also lead us to namespacing them, which could eliminate the 'restart' arg.)

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+        return self._name
+
+
+class RunWithLock(RollingEvent):
```
This and the other event classes should probably have some nice docstrings :)

pengale replied:
Agreed.

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+class RollingOpsManager(Object):
+    """Emitters and handlers for rolling ops."""
+
+    def __init__(self, charm, rel_name, func_run, event_acquire):
```
So I dropped a comment up top about this, but I think we should consider the following signature here:

```python
def __init__(self, charm, *, relation, trigger, callback):
```

To me at least, the language `trigger` and `callback` more obviously conveys the intent here? Open to discussion, just my two cents! :)

pengale replied:
That makes sense. I think that I was trying to avoid "callback" because this is almost, but not quite a callback. But it reads a lot cleaner, that way.

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+        super().__init__(charm, None)
+
+        self.charm = charm
+        self.name = rel_name
```
Might it be worth doing a rudimentary check against the charm's metadata to ensure that the relation name specified exists, and is a peer relation?

jnsgruk commented on lib/charms/rolling_ops/v0/rollingops.py:
```diff
@@ -0,0 +1,353 @@
+    pass
+
+
+class RollingEvents(CharmEvents):
```
Could there be merit in having a `RollingComplete` (name to be discussed...!) event that is emitted once *all* units have held the lock and run the callback? Such an event could be useful for charm authors to be able to subscribe to for clearing up statuses in the model?

pengale replied:
We could. But then we start getting into the territory where we really want timestamps, so that a rolling complete event doesn't get us into a state where we can't fire the next rolling restart.

jnsgruk reviewed (commented):
Overall this is looking nice! Thanks Pen!

pengale commented:
Thanks for the comments, all. I've been busy hammering at this on a live Juju instance, and haven't looped back to update the PR based on the comments. But I have read and appreciated all of them, and should have something updated to push soon :-)

Commit:
Ran linter and addressed smaller PR comments.

Cleaned up lint failures.

Addressed multiple PR comments, including fixing up naming, and moving to enums for lock states.

Changed "status" of lock to "state" of lock (to disambiguate from unit status).

Moved binding of restart_action out of the library, in preparation for
suggesting that folks just do "juju run"

Added an optional "delay" to restart action.

Still working on getting everything working properly in live environment.

Commit:
Namespaced events, and fixed up inheritance.

Refactored the RollingOpsManager so that it defines a series of
namespaced events. This allows us to auto attach them, without
worrying too much about clobbering.

Also simplified the way that we're passing Framework state around, to
rely more closely on the Object meta class.

Fixed statuses. We now set an application wide status, which the smoke
test watches for.

pengale commented:
Merged the code. Going to take the points still under discussion into separate issues/PRs.
